### PR TITLE
[processor/interval] Support Gauges and Summaries

### PR DESCRIPTION
**Pull request by ArthurSens**
**Description:** Adds support for Gauges and Summaries 

**Link to tracking Issue:** #34803 

**Testing:** Unit tests were extended to cover the new behavior

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Review by jpkrohling — COMMENTED** [review]
I'll wait for @sh0rez's review as well before merging, and apart from a typo (fufill), this LGTM.

**Comment by RichieSams** [comment]
I think the implementation is good. I'm still a bit undecided on whether we *should* aggregate gauges and summaries though.

With monotonically increasing sums and histograms, by aggregating you're losing precision, but you're not losing data. But for gauges, summaries, and non-monotonically increasing sums, if you only report the "last" value, you're losing data.

Maybe this is fine. But if that's the case, then we should update the docs to highlight that fact. So if people *don't* want that behavior, they'll have to set up routing rules to not send gauges, etc to this processor.

**Comment by ArthurSens** [comment]
Good point, I'm coming here with the objective of reducing data throughput and from a Prometheus background. In the Prometheus world you only care about the value during the scrape, but I can see how in a push model you might want to actively push gauges in a specific situation and not lose this information.


Would it make sense to make it a configuration option? The user could decide if gauges/summaries must pass through or only the last value should be kept.

File: processor/intervalprocessor/processor.go
Comment (sh0rez) [inline processor/intervalprocessor/processor.go:111]: it in fact does! all the required methods are there, but the interface is overly restrictive because it has a concrete type list.
it appears to serve no real purpose, because aggregateDataPoints can operate on any type carrying those methods.

I'd suggest removing the type list entirely, making the interface purely behavioral:

```diff
--- a/processor/intervalprocessor/internal/metrics/metrics.go
+++ b/processor/intervalprocessor/internal/metrics/metrics.go
 type DataPoint[Self any] interface {
-       pmetric.NumberDataPoint | pmetric.HistogramDataPoint | pmetric.ExponentialHistogramDataPoint

        Timestamp() pcommon.Timestamp
        Attributes() pcommon.Map
        CopyTo(dest Self)
}
```

**Review by tombrk — CHANGES_REQUESTED** [review]

**Comment by sh0rez** [comment]
> @RichieSams Maybe this is fine. But if that's the case, then we should update the docs to highlight that fact. So if people don't want that behavior, they'll have to set up routing rules to not send gauges, etc to this processor.

I think it is. My general expectation of this processor would be "sends all data at the specified interval", which is really relevant for cloud platforms with datapoint/minute based billing. 

> @ArthurSens Would it make sense to make it a configuration option? The user could decide if gauges/summaries must pass through or only the last value should be kept.

I think its common enough and rarely does harm if enabled by default. _However_ I think we should provide an easy way to opt out, e.g. `lossy: false`. While you could use filterprocessor and setup two pipelines to selectively route to this component, this is more complex and I think that extra config option doesn't cost us a lot. 

---

Docs idea:

```md
Changing the data interval is an inherently lossy operation.
For some datatypes this only results in slight precision loss, while others will lose actual observations.
While the latter is generally deemed acceptable, one may opt-out by setting `lossy: false`

| | Datatype |
|-|-|
| :information_source: precision loss | Monotonic Sum, Histogram, Exponential Histogram |
| :warning: data loss | Gauge, Summary, Non-monotonic Sum |
```

File: processor/intervalprocessor/processor.go
Comment (RichieSams) [reply processor/intervalprocessor/processor.go:111]: Yea, I was being overly restrictive. I'm ok with removing the hard type requirements

**Comment by RichieSams** [comment]
I like @sh0rez 's approach! I think we may want to further clarify what `lossy: false` does though: IE, all "data loss" potential types are passed through. Whereas "precision loss" are still accumulated.

File: processor/intervalprocessor/processor.go
Comment (ArthurSens) [reply processor/intervalprocessor/processor.go]: done :)

**Comment by ArthurSens** [comment]
I hard a difficult time explaining in docs/comments the behavior while using `lossy`, so I suggested a different configuration option 🙈. What do you think?

File: processor/intervalprocessor/README.md
Comment (RichieSams) [inline processor/intervalprocessor/README.md:30]: Maybe:

```
NOTE: Aggregating data over an interval is an inherently "lossy" process. For monotonically increasing, cumulative sums, histograms, and exponential histograms, you "lose" precision, but you don't lose overall data. But for non-monotonically increasing sums, gauges, and summaries, aggregation represents actual data loss. IE you could "lose" that a value increased and then decreased back to the original value. In most cases, this data "loss" is ok. However, if you would rather these values be passed through, and *not* aggregated, you can set that in the configuration
```


Thoughts?

File: processor/intervalprocessor/README.md
Comment (RichieSams) [inline processor/intervalprocessor/README.md:38]: Should we add one for non-monotonic sums? IMO they're no different than gauges.

File: processor/intervalprocessor/README.md
Comment (ArthurSens) [reply processor/intervalprocessor/README.md:30]: Much better

File: processor/intervalprocessor/README.md
Comment (RichieSams) [reply processor/intervalprocessor/README.md:38]: Fine by me

Comment (sh0rez) [reply processor/intervalprocessor/README.md:38]: nit:

```yaml
ignore:
  gauges: true
  summaries: true
```

feels more elegant, but no strong opinion at all

**Review by tombrk — APPROVED** [review]

File: processor/intervalprocessor/testdata/summaries_are_aggregated/input.yaml
Comment (jpkrohling) [inline processor/intervalprocessor/testdata/summaries_are_aggregated/input.yaml:2]: I have a feeling that this is copied from some existing files, _but_ I wanted to mention that `example.com` is the domain to be used in tests as per RFC 2606.

**Review by jpkrohling — APPROVED** [review]

File: processor/intervalprocessor/testdata/summaries_are_aggregated/input.yaml
Comment (jpkrohling) [reply processor/intervalprocessor/testdata/summaries_are_aggregated/input.yaml:2]: I'm merging as is, but if you feel like fixing this, open a follow-up PR.

File: processor/intervalprocessor/README.md
Comment (RichieSams) [reply processor/intervalprocessor/README.md:38]: I like @sh0rez 's idea. Though, I would go for `pass_through` instead of `ignore`. To be a bit more explicit

**Pull request merged**
